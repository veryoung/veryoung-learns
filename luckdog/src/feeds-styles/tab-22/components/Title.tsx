// UI8的title信息，包括title和右边的快链
import React from 'react';
import {
  StyleSheet,
  View,
  Image,
  Text,
} from '@tencent/hippy-react-qb';
import { FeedsUIStyle, FeedsLineHeight, FeedsTheme } from './utils';
import FeedsIcon from '../../../framework/FeedsIcon';
import { shouldComponentUpdate, countReRender } from '@tencent/luckbox-react-optimize';
import { getWidthHeight } from '../../../framework/utils/device';
import { strictExposeReporter } from '@/luckdog';
import { CLICK_STEP } from '../../../framework/FeedsConst';
import { throttle } from '@/luckbox';
import { CommonProps, RightLink } from '../../../entity';

const windowWidth = getWidthHeight().width;
const titleMaxWidth = ((windowWidth - 24) / 4) * 3;

const styles = StyleSheet.create({
  arrow: {
    height: 9,
    width: 5,
    marginLeft: 4,
    tintColors: FeedsTheme.SkinColor.N1,
  },
  box: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: FeedsUIStyle.FEEDS_CARD_MARGIN_VERTICAL,
  },
  moreStr: {
    colors: FeedsTheme.SkinColor.N1,
    fontSize: FeedsUIStyle.T1,
    lineHeight: FeedsLineHeight.T1,
  },
  moreView: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'flex-end',
    paddingBottom: FeedsUIStyle.FEED_TITLE_HEIGHT,
    paddingTop: FeedsUIStyle.FEED_TITLE_HEIGHT,
    position: 'relative',
  },
  redDot: {
    height: 5,
    width: 5,
    borderRadius: 2.5,
    backgroundColors: FeedsTheme.LiteColor.B2,
    marginRight: 5,
    alignItems: 'center',
    justifyContent: 'center',
  },
  txt: {
    colors: FeedsTheme.SkinColor.A1,
    fontSize: FeedsUIStyle.T3,
    fontWeight: 'bold',
    lineHeight: FeedsLineHeight.T3,
    paddingBottom: FeedsUIStyle.FEED_TITLE_HEIGHT,
    paddingTop: FeedsUIStyle.FEED_TITLE_HEIGHT,
    maxWidth: titleMaxWidth,
  },
});

interface TitleProps extends CommonProps {
  /** 标题 */
  title?: string;
  /** 右上角跳转 */
  right?: RightLink;
  /** 是否展示换一换 */
  changeable?: boolean;
  /** 是否显示红点 */
  showDot?: boolean;
  /** 右上角跳转自定义样式 */
  moreTextStyle?: Record<string, any>;
  /** 标题自定义样式 */
  titleStyle?: Record<string, any>;
  /** 点击上报 */
  doBeaconByClick?: () => void;
  /** 点击换一换 */
  switchNovel?: () => void;
  /** 点击右上角事件 */
  rightClick?: () => void;
  /** 不需要补报曝光 */
  noExposeAddition?: boolean;
}

export class Title extends React.Component {
  public props!: TitleProps;
  public state: {
    hasClick: boolean;
  };

  public handleClick = throttle(() => {
    const { parent = {}, right, doBeaconByClick } = this.props;
    this.setState({ hasClick: true });
    doBeaconByClick?.();
    parent.loadUrl?.(right?.sUrl);
  }, CLICK_STEP);

  public constructor(props: any) {
    super(props);
    this.shouldComponentUpdate = shouldComponentUpdate(this, 'Title');
    this.state = {
      hasClick: false,
    };
  }

  public handleClickChange = () => {
    const { doBeaconByClick, switchNovel } = this.props;
    doBeaconByClick?.();
    switchNovel?.();
  };

  /**
 * 展示：假使用户我的收藏书籍中有任一书籍保持在更新状态且不外露在首屏，则查看全部红点会展示
   消失：假使用户我的收藏书籍中所有书籍都不再更新状态，或主动点击了查看全部红点后，查看全部红点会消失
 */
  public render() {
    countReRender(this, 'Title');
    const {
      title = '',
      right,
      changeable,
      showDot,
      moreTextStyle = {},
      titleStyle = {},
      rightClick,
      noExposeAddition,
    } = this.props || {};
    let link;
    const { hasClick } = this.state;
    const rightClickHandler = rightClick ? rightClick : this.handleClick;

    if (right?.sUrl) {
      let redDot;
      if (showDot && !hasClick) {
        redDot = <View style={styles.redDot} />;
      }
      link = (
        <View
          className='right-view'
          style={styles.moreView}
          onClick={noExposeAddition ? rightClickHandler : strictExposeReporter.triggerExpoCheck(rightClickHandler)}
        >
          {redDot}
          <Text style={[styles.moreStr, moreTextStyle]} numberOfLines={1}>
            {right.sText}
          </Text>
          <Image
            source={{ uri: FeedsIcon.novel_card_arrow }}
            style={styles.arrow}
            noPicMode={{ enable: false }}
          />
        </View>
      );
    } else if (changeable) {
      link = (
        <View className='change-view' style={styles.moreView} onClick={strictExposeReporter.triggerExpoCheck(this.handleClickChange)}>
          <Text style={[styles.moreStr, moreTextStyle]} numberOfLines={1}>
            换一换
          </Text>
        </View>
      );
    }
    return (
      <View
        style={styles.box}
        onLayout={event => this.props.onLayout?.(event)}
      >
        <Text style={[styles.txt, titleStyle]} numberOfLines={1}>
          {title}
        </Text>
        {link}
      </View>
    );
  }
}
