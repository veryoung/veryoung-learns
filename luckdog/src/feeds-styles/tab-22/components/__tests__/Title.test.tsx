import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import { Title } from '../Title';
import { commonProps } from '@test/mocks/props';

/**
 * @author azeryang
 * @priority P0
 * @casetype unit
 */
describe('<Title />', () => {
  describe('test render', () => {
    it('with title', () => {
      const props = {
        ...commonProps,
        title: '标题',
      };
      const tree = renderer.create(<Title {...props} />);
      expect(tree).toMatchSnapshot();
    });

    it('with changeable（换一换）', () => {
      const props = {
        ...commonProps,
        title: '标题',
        changeable: true,
      };
      const tree = renderer.create(<Title {...props} />);
      expect(tree).toMatchSnapshot();
    });

    it('with right（右上角链接 + 红点）', () => {
      const props = {
        ...commonProps,
        right: {
          sUrl: 'https://novel.html5.qq.com/qbread/intro?bookid=1100480477',
          sText: '简介页',
        },
        showDot: true,
      };
      const tree = renderer.create(<Title {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });

  describe('test click', () => {
    it('change click（换一换点击）', () => {
      const props = {
        ...commonProps,
        changeable: true,
        doBeaconByClick: jest.fn(),
        switchNovel: jest.fn(),
      };
      const wrap = shallow(<Title {...props} />);
      wrap.find('.change-view').simulate('click');
      expect(props.doBeaconByClick.mock.calls.length).toBe(1);
      expect(props.switchNovel.mock.calls.length).toBe(1);
    });

    it('right click（右上角点击）', () => {
      const props = {
        ...commonProps,
        right: {
          sUrl: 'https://novel.html5.qq.com/qbread/intro?bookid=1100480477',
          sText: '简介页',
        },
        doBeaconByClick: jest.fn(),
        showDot: true,
      };
      const wrap = shallow(<Title {...props} />);
      expect(wrap.state('hasClick')).toBeFalsy();
      wrap.find('.right-view').simulate('click');
      expect(wrap.state('hasClick')).toBeTruthy();
      expect(props.doBeaconByClick.mock.calls.length).toBe(1);
    });
  });
});
