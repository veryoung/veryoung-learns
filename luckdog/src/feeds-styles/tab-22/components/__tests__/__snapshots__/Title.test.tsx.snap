// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Title /> test render with changeable（换一换） 1`] = `
<div
  nativeName="View"
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 16,
    }
  }
>
  <p
    nativeName="Text"
    numberOfLines={1}
    style={
      Array [
        Object {
          "colors": Array [
            "#242424",
            "#686d74",
            "#242424",
            "white",
          ],
          "fontSize": 16,
          "fontWeight": "bold",
          "lineHeight": 20,
          "maxWidth": 252,
          "paddingBottom": 14,
          "paddingTop": 14,
        },
        Object {},
      ]
    }
    text="标题"
  >
    标题
  </p>
  <div
    className="change-view"
    nativeName="View"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "paddingBottom": 14,
        "paddingTop": 14,
        "position": "relative",
      }
    }
  >
    <p
      nativeName="Text"
      numberOfLines={1}
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 12,
            "lineHeight": 16,
          },
          Object {},
        ]
      }
      text="换一换"
    >
      换一换
    </p>
  </div>
</div>
`;

exports[`<Title /> test render with right（右上角链接 + 红点） 1`] = `
<div
  nativeName="View"
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 16,
    }
  }
>
  <p
    nativeName="Text"
    numberOfLines={1}
    style={
      Array [
        Object {
          "colors": Array [
            "#242424",
            "#686d74",
            "#242424",
            "white",
          ],
          "fontSize": 16,
          "fontWeight": "bold",
          "lineHeight": 20,
          "maxWidth": 252,
          "paddingBottom": 14,
          "paddingTop": 14,
        },
        Object {},
      ]
    }
    text=""
  >
    
  </p>
  <div
    className="right-view"
    nativeName="View"
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "paddingBottom": 14,
        "paddingTop": 14,
        "position": "relative",
      }
    }
  >
    <div
      nativeName="View"
      style={
        Object {
          "alignItems": "center",
          "backgroundColors": Array [
            "#f55318",
            "#74322d",
          ],
          "borderRadius": 2.5,
          "height": 5,
          "justifyContent": "center",
          "marginRight": 5,
          "width": 5,
        }
      }
    />
    <p
      nativeName="Text"
      numberOfLines={1}
      style={
        Array [
          Object {
            "colors": Array [
              "#000000",
              "#686D74",
              "#000000",
              "#ffffff",
            ],
            "fontSize": 12,
            "lineHeight": 16,
          },
          Object {},
        ]
      }
      text="简介页"
    >
      简介页
    </p>
    <img
      alt=""
      nativeName="Image"
      noPicMode={
        Object {
          "enable": false,
        }
      }
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA8AAAAbCAYAAACjkdXHAAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAD6ADAAQAAAABAAAAGwAAAACSVupLAAAA40lEQVQ4EZ1UOQ4CMQyMQDQ8iIqWJ3HDAhXiEJf2KZS0PAX2B7QwRrFkuZo40sheecZJ1nZSSmkDNMACKF5vKL4Z2xJ1G+QWMMiiPmwXeORvysjRdXexe0plSGv4NsHBxCi3cgmOlMqQli7BycQoV8pmr3CmVIY0cwkuJka5U5fgSqkMaeIS3EyMcscuQS0q6TBmPUH6ANqJPfgdRmg5d3xoFWSY6DUEU4Vi6SEaOeH/zsy24b8drnO4w+bujnRvh6cqPM8rd1T6JQm/Yb4cO6b4ypEe1bajW07F8uS8AOmkovUDoSNiNZ1r/OEAAAAASUVORK5CYII="
      style={
        Object {
          "height": 9,
          "marginLeft": 4,
          "tintColors": Array [
            "#000000",
            "#686D74",
            "#000000",
            "#ffffff",
          ],
          "width": 5,
        }
      }
    />
  </div>
</div>
`;

exports[`<Title /> test render with title 1`] = `
<div
  nativeName="View"
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 16,
    }
  }
>
  <p
    nativeName="Text"
    numberOfLines={1}
    style={
      Array [
        Object {
          "colors": Array [
            "#242424",
            "#686d74",
            "#242424",
            "white",
          ],
          "fontSize": 16,
          "fontWeight": "bold",
          "lineHeight": 20,
          "maxWidth": 252,
          "paddingBottom": 14,
          "paddingTop": 14,
        },
        Object {},
      ]
    }
    text="标题"
  >
    标题
  </p>
</div>
`;
